### PR TITLE
LibPQ: Checks for NUL bytes before sending value to LibPQ

### PR DESCRIPTION
**Pull request by qxjit**
This adds a new type, `PGTextFormatValu,e` to track whether we know that
the bytes it contains are all non-NUL (e.g. because we constructed
directly). For any values we do _not_ know are non-NUL, we check them at
the time of building the param list to pass to lib and abort running
the query. For now `executeRow` return `Nothing`, the same as if LibPQ
returned an error. We have another story to start doing better error
handling in general, so I have left actually raising the error about
NUL bytes for that story.

### CODE REVIEW

File: orville-postgresql-libpq/src/Database/Orville/PostgreSQL/Internal/PGTextFormatValue.hs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  Note that postgresql does not allow NUL bytes in text values, and the LibPQ C
+  library expects text values to be given as NULL-terminated C Strings, so
+  '\NUL' bytes cannot be included in a 'PGTextFormatValue'. If 'fromByteString'
+  is used to construct the 'PGTextFormatValue' (normally what you should do),
+  an error will be raised before libpq is called to execute the query. If
+  'unsafeFromByteString' is used, the caller is expected to ensure that no
+  '\NUL' bytes are present. If a '\NUL' byte is included with
+  'unsafeFromByteString', the value passed to the database will be truncated at
+  the '\NUL' byte because it will be interprested as the end of the C String by
```
Comment (ysangkok) [inline orville-postgresql-libpq/src/Database/Orville/PostgreSQL/Internal/PGTextFormatValue.hs:26]: s/interprested/interpreted/

File: orville-postgresql-libpq/src/Database/Orville/PostgreSQL/Internal/PGTextFormatValue.hs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  libpq.
+-}
+data PGTextFormatValue
+  = NoAssumptionsMade BS.ByteString
+  | AssumedToHaveNoNULValues BS.ByteString
```
Comment (ysangkok) [inline orville-postgresql-libpq/src/Database/Orville/PostgreSQL/Internal/PGTextFormatValue.hs:31]: Pretty much any statement in source code can be invalidated by the presence of a bug, so what about something simpler like "MayContainNUL" and "WithoutNULs"? But I don't feel strongly, and I don't know the code well, feel free to ignore. Just my two cents :)

**Review by ysangkok — COMMENTED** [review]
Dunno if this is too much bikeshedding :P

File: orville-postgresql-libpq/src/Database/Orville/PostgreSQL/Internal/PGTextFormatValue.hs (hunk 2 above)
Comment (qxjit) [reply orville-postgresql-libpq/src/Database/Orville/PostgreSQL/Internal/PGTextFormatValue.hs:31]: The reason I didn't chose something `WithoutNULs` is that it could easily imply to a reader that the value is _guaranteed_ to have no nulls because it was checked at runtime, which is not the case. As far as the `PGTextFormatValue` module is concerned it is purely an assumption that there are no nulls. Given that the constructors are not exported from the module anyhow, I erred on the side of precision of meaning, even though it may be a bit wordy.

Basically, if someone ends up needing to muck around in the `PGTextFormatValue` implementation, I want to surface that fact that there is no enforcement regarding nulls in those values to them as soon as possible.
